python/dgl/backend/mxnet/tensor.py

```diff
@@ -368,28 +369,58 @@ def forward(self, lhs_data, rhs_data):
                             ctx=lhs_data.context, dtype=lhs_data.dtype)
         out_data_nd = zerocopy_to_dgl_ndarray_for_write(out_data)
         K.binary_op_reduce(
-            self.reducer, self.binary_op, self.graph, self.lhs, self.rhs,
+            self.reducer if self.reducer != 'mean' else 'sum',
+            self.binary_op, self.graph, self.lhs, self.rhs,
             lhs_data_nd, rhs_data_nd, out_data_nd, self.lhs_map[0],
             self.rhs_map[0], self.out_map[0])
+        # normalize if mean reducer
+        # NOTE(zihao): this is a temporary hack and we should have better solution in the future.
+        if self.reducer == 'mean':
+            degs = nd.empty((out_data.shape[0],),
+                            ctx=out_data.context, dtype=out_data.dtype)
+            degs_nd = zerocopy_to_dgl_ndarray(degs)
+            if self.lhs != TargetCode.DST: # copy_reduce does not accept dst node as input target.
+                in_ones = nd.ones((lhs_data.shape[0],),
+                                  ctx=lhs_data.context, dtype=lhs_data.dtype)
+                in_ones_nd = zerocopy_to_dgl_ndarray(in_ones)
+                K.copy_reduce(
+                    'sum', self.graph, self.lhs, in_ones_nd, degs_nd, 
+                    self.lhs_map[0], self.lhs_map[0])
+            else: # self.rhs != TargetCode.DST
+                in_ones = nd.ones((rhs_data.shape[0],),
+                                  ctx=rhs_data.context, dtype=rhs_data.dtype)
+                in_ones_nd = zerocopy_to_dgl_ndarray(in_ones)
+                K.copy_reduce(
+                    'sum', self.graph, self.rhs, in_ones_nd, degs_nd, 
+                    self.rhs_map[0], self.rhs_map[0])
```

Comment:
Does the following work?
```python
degs = nd.empty((out_data.shape[0],), ctx=out_data.context, dtype=out_data.dtype)
degs_nd = zerocopy_to_dgl_ndarray(degs)
in_ones = nd.ones((graph.number_of_nodes(),), ctx=lhs_data.context, dtype=lhs_data.dtype)
in_ones_nd = zerocopy_to_dgl_ndarray(in_ones)
K.copy_reduce('sum', self.graph, TargetCode.SRC, in_ones_nd, degs_nd)
```

The mapping arrays can be omitted in your case.

Reply:
Yes, it works. But I was wondering why could we not use mapping information? What if we update a subgraph instead of the whole graph?

Reply:
btw, by calling `g.send_and_recv(eids, msg_func, reduce_func)` I met another bug:
```
  File "/home/ec2-user/dgl/python/dgl/graph.py", line 2523, in send_and_recv
    inplace=inplace)
  File "/home/ec2-user/dgl/python/dgl/runtime/scheduler.py", line 183, in schedule_snr
    out_map_creator=out_map_creator)
  File "/home/ec2-user/dgl/python/dgl/runtime/scheduler.py", line 859, in _gen_send_reduce
    adj, edge_map, nbits = adj_creator()
  File "/home/ec2-user/dgl/python/dgl/runtime/scheduler.py", line 171, in <lambda>
    edge_tuples, graph._number_of_src_nodes(), graph._number_of_dst_nodes())
  File "/home/ec2-user/dgl/python/dgl/runtime/spmv.py", line 181, in build_gidx_and_mapping_uv
    gidx = create_bipartite_from_coo(num_src, num_dst, u, v)
  File "/home/ec2-user/dgl/python/dgl/heterograph_index.py", line 799, in create_bipartite_from_coo
    return _CAPI_DGLHeteroCreateBipartiteFromCOO(
NameError: name '_CAPI_DGLHeteroCreateBipartiteFromCOO' is not defined
```

Reply:
I'm afraid not using mapping is problematic, for example:
```python
import dgl
import dgl.function as fn
import torch as th
g = dgl.DGLGraph()
g.add_nodes(4)
g.add_edges([2, 0, 0, 1, 1, 0], [2, 1, 2, 1, 2, 0])
feat = th.arange(1, 5).float().unsqueeze(-1).cuda()
feat.requires_grad = True
g.ndata['x'] = feat
g.edata['x'] = th.ones(6, 1).cuda()
g.send_and_recv(th.LongTensor([0, 2, 4, 5]), fn.u_mul_e('x', 'x', 'm'), fn.mean('m', 'o'))
print(g.ndata['o'])
g.ndata['o'].sum().backward()
print(g.ndata['x'].grad)
```
With mapping information, the output is:
```
tensor([[1.],
        [0.],
        [2.],
        [0.]], device='cuda:0', grad_fn=<IndexCopyBackward>)
tensor([[1.3333],
        [0.3333],
        [0.3333],
        [0.0000]], device='cuda:0')
```
Without mapping information, the output is:
```python
tensor([[1.],
        [0.],
        [6.],
        [0.]], device='cuda:0', grad_fn=<IndexCopyBackward>)
tensor([[2.],
        [1.],
        [1.],
        [0.]], device='cuda:0')
```

Reply:
`K.copy_reduce('sum', self.graph, TargetCode.SRC, in_ones_nd, degs_nd, None, self.out_map[0])` would solve the problem.

Reply:
@jermainewang When doing `send_and_recv`, `self.graph` here is a bipartite graph index, and using `graph.number_of_nodes()` can be problematic. Here if we use `copy_reduce(target=src)`, we actually need `number_of_nodes` to return number of source nodes.

I think an alternative is to use `copy_edge` for computing node degree. And then the one tensor can be created with `nd.ones((graph.number_of_edges(),))`.

However, the problem is the bipartite graph interface is slightly different. I checked the C++ implementation of bipartite graph, and found that it subclasses heterograph and therefore its API needs to take in an argument `etype`. For example:
https://github.com/dmlc/dgl/blob/73b2668fe260017bb09e74553b65dc4a6094afa6/src/graph/bipartite.cc#L109-L111
The `etype` argument was not really used, but it makes the bipartite graph index interface different from ordinary graph index. @jermainewang @BarclayII I wonder why bipartite graph is designed this way. Is there a reason bipartite graph needs to handle more than one edge type?

Reply:
The bipartite graph interface is not stable enough for use. Let's not add extra complexity here. For @ylfdq1118 's question, etype here is simply 0. A check could be enforced too. We will gradually move all the kernels to general sparse matrix so we will definitely come back to this later.